feat: improve openfga connection handling

kian99 opened the pull request:
## Description

This PR started with a similar reason as https://github.com/canonical/jimm-k8s-operator/pull/81, where JIMM would always use stale data from its peer relation instead of fetching the latest connection details from the app relation (see the linked issue for more details).

In addition to the above, the JIMM charm was creating a hash of the desired OpenFGA auth model and comparing that with the value in its state. If the two matched, JIMM would not recreate the auth model if OpenFGA, expecting that if it exists in the charm's state, it must exist in OpenFGA. But in the current JIMM deployment, the database was recreated so the auth model no longer exists in OpenFGA. This PR adjusts JIMM's logic to rather check OpenFGA instead of relying on its own permanent cache.

Fixes [JUJU-8386](https://warthogs.atlassian.net/browse/JUJU-8386)

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

[JUJU-8386]: https://warthogs.atlassian.net/browse/JUJU-8386?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kian99 committed:
feat: improve openfga auth model creation logic

kian99 commented on src/charm.py:
```diff
@@ -913,13 +904,19 @@ def _update_trusted_ca_certs(self, container: Container) -> bool:
                         certs.append([cert_path, cert])
             # set certs in peer relation databag, if they are changed
             if certs:
-                existing_certs_secret = json.loads(self.model.get_relation("peer").data[self.app].get("ca", "[]"))
```
The changes from here and below were just fixing some linter warnings that the value could be `None`.

kian99 replied:
Reverted this in case it's causing the issues with the integration test where it look like there is a hook loop (some hook triggering the same hook) and keeping the JIMM unit's in an "executing" state forever.

luci1900 reviewed (approved):
Looks good.

Continuing the trend of charm internals scaring me, the hashing certainly does that. Glad to see it gone.

SimoneDutto commented on src/charm.py:
```diff
@@ -806,69 +804,62 @@ def setup_fga_auth_model(self, jimm_container: Container) -> None:
+                logger.error("failed to fetch existing authorization model: %s", e)
+                logger.warning("skipping auth model creation")
+                return
+            remote_model = remote.get("authorization_model", remote) if remote else None
```
if remote can be `None` after the try expect i would rather return early, than using this if else

SimoneDutto replied:
in this way you can simplify the two `is_` conditions

kian99 replied:
Yup, will simplify.

SimoneDutto commented on src/charm.py:
```diff
@@ -806,69 +804,62 @@ def setup_fga_auth_model(self, jimm_container: Container) -> None:
         if dns is None or dns == "":
             dns = "http://localhost"
-        dns = ensureFQDN(dns)
+        dns = ensureFQDN(str(dns))
```
why this cast?

kian99 replied:
According to my type hints, `self.config.get("dns-name")` can return `(bool | int | float | str | None)` so this was the easiest way to simplify the type hint warning there.

SimoneDutto reviewed (commented):
some comments

kian99 committed:
refactor: improve setup_fga_auth_model logic

kian99 committed:
refactor: update checks for auth model equality

kian99 committed:
feat: store openfga model md5 digest

The OpenFGA server doesn't return the auth model in the same format that it was supplied so we need to store and compare a hash of the model on the charm.
A relevant issue on the OpenFGA Github has been linked.
